File: Question 0450 - Delete Node in a BST/src/Solution.java
class Solution {

    private TreeNode minVal(TreeNode root)
    {
        TreeNode temp = root;
        while (temp.left != null)
            temp = temp.left;

        return temp;
    }
    public TreeNode deleteNode(TreeNode root, int key)
    {
        //base case
        if(root == null) return root;

        if(root.val == key)
        {
            //0 child
            if(root.left == null && root.right == null)
            {
                root = null;
                return null;
            }

            //1 child ->left

            if(root.left != null && root.right == null)
            {
                TreeNode temp = root.left;
                root = null;
                return temp;
            }

            //right child
            if(root.left == null && root.right != null)
            {
                TreeNode temp = root.right;
                root = null;
                return temp;
            }

            //2 child
            if(root.left != null && root.right != null)
            {
                int min = minVal(root.right).val;

                root.val = min;
                root.right = deleteNode(root.right, min);
                return root;
            }
        }

        else if(root.val > key)
        {
            root.left = deleteNode(root.left, key);
            return root;
        }
        else
        {
            root.right = deleteNode(root.right, key);
            return root;
        }

        return root;
    }
}

File: Question 0450 - Delete Node in a BST/src/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
    TreeNode *minVal(TreeNode *root)
    {
        TreeNode *tmp = root;
        while (tmp->left)
            tmp = tmp->left;

        return tmp;
    }

public:
    TreeNode *deleteNode(TreeNode *root, int key)
    {
        if (!root)
            return root;

        if (root->val == key)
        {

            // 0 child
            if (!root->left && !root->right)
            {
                root = NULL;
                return NULL;
            }

            // 1 child
            if (root->left && !root->right)
            {
                TreeNode *tmp = root->left;
                root = NULL;
                return tmp;
            }

            if (!root->left && root->right)
            {
                TreeNode *tmp = root->right;
                root = NULL;
                return tmp;
            }

            // 2 children
            if (root->left && root->right)
            {
                int mini = minVal(root->right)->val;

                root->val = mini;
                root->right = deleteNode(root->right, mini);
                return root;
            }
        }
        else if (root->val > key)
        {
            root->left = deleteNode(root->left, key);
            return root;
        }
        else
        {
            root->right = deleteNode(root->right, key);
            return root;
        }
    }
};